crates/miden-objects/src/block/account_tree.rs

```diff
@@ -0,0 +1,432 @@
+/// [`AccountId::prefix`]. In other words, account ID prefixes are unique in the blockchain.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AccountTree {
+    smt: Smt,
+}
+
+impl AccountTree {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// The depth of the account tree.
+    pub const DEPTH: u8 = SMT_DEPTH;
+
+    /// The index of the account ID suffix in the SMT key.
+    pub(super) const KEY_SUFFIX_IDX: usize = 2;
+    /// The index of the account ID prefix in the SMT key.
+    pub(super) const KEY_PREFIX_IDX: usize = 3;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new, empty account tree.
+    pub fn new() -> Self {
+        AccountTree { smt: Smt::new() }
+    }
+
+    /// Returns a new [`Smt`] instantiated with the provided entries.
+    ///
+    /// If the `concurrent` feature of `miden-crypto` is enabled, this function uses a parallel
+    /// implementation to process the entries efficiently, otherwise it defaults to the
+    /// sequential implementation.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the provided entries contain multiple commitments for the same account ID.
+    /// - multiple account IDs share the same prefix.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (AccountId, Digest)>,
+    ) -> Result<Self, AccountTreeError> {
+        let smt = Smt::with_entries(
+            entries
+                .into_iter()
+                .map(|(id, commitment)| (Self::id_to_smt_key(id), Word::from(commitment))),
+        )
+        .map_err(|err| {
+            let MerkleError::DuplicateValuesForIndex(leaf_idx) = err else {
+                unreachable!("the only error returned by Smt::with_entries is of this type");
+            };
+
+            // SAFETY: Since we only inserted account IDs into the SMT, it is guaranteed that
+            // the leaf_idx is a valid Felt as well as a valid account ID prefix.
+            AccountTreeError::DuplicateStateCommitments {
+                prefix: AccountIdPrefix::new_unchecked(
+                    Felt::try_from(leaf_idx).expect("leaf index should be a valid felt"),
+                ),
+            }
+        })?;
+
+        for (leaf_idx, leaf) in smt.leaves() {
+            if leaf.num_entries() >= 2 {
+                // SAFETY: Since we only inserted account IDs into the SMT, it is guaranteed that
+                // the leaf_idx is a valid Felt as well as a valid account ID prefix.
+                return Err(AccountTreeError::DuplicateIdPrefix {
+                    duplicate_prefix: AccountIdPrefix::new_unchecked(
+                        Felt::try_from(leaf_idx.value())
+                            .expect("leaf index should be a valid felt"),
+                    ),
+                });
+            }
+        }
```

Comment:
We could check if the number of leaves is different from the number of passed in accounts before iterating over the leaves. This way, we'd iterate over leaves only if we know there was an issue.

Let's make this change in a follow-up PR.